<h1 style="color: #001a79;">Randomness</h1>

<h3 style="color: #001a79;">Exercise 1</h3>

<hr style="border-top: 1px solid #001a79;" />

It is somewhat interesting that `(5 * 4 * 3 * 2 * 1)` perfectly divides `(10 * 9 * 8 * 7 * 6)` - there's no remainder.

If we only wanted exactly four heads as opposed to five, the equivalent calculation would be `(10 * 9 * 8 * 7) / (4 * 3 * 2 * 1)`.

Does that evenly divide too? What is the formula in general?

Does it always come out as a positive whole number?

In [1]:
numerator = (10 * 9 * 8 * 7)
numerator

5040

In [2]:
denominator = (4 * 3 * 2 * 1)
denominator

24

In [3]:
numerator // denominator

210

This divides evenly. 

The general formula is: $${n \choose r} = {n! \over r!(n-r)!}$$

where n is the number of things to choose from, and we choose r of them. 

https://www.mathsisfun.com/combinatorics/combinations-permutations.html

If we use the above equation to solve $${10 \choose 4} = {10! \over 4!(10-4)!} = {10! \over (4!)(6!)} =$$

$${(10\times 9\times 8\times 7\times 6\times 5\times 4\times 3\times 2\times 1)} \over {(4\times 3\times 2\times 1)(6\times 5\times 4\times 3\times 2\times 1)}$$

$${(10\times 9\times 8\times 7)} \over {(4\times 3\times 2\times 1)}$$

As above we see we get a total of 210 combinations choosing 4 from 10. 

In [4]:
import math
numerator = (10 * 9 * 8 * 7)
denominator = (4 * 3 * 2 * 1)
numerator // denominator

210

Does it always come out as a positive whole number? 

In short yes. 

The below triangle is named Pascal's Triangle (named after Blaise Pascal, a famous French Mathematician and Philosopher).

![Pascals Triangle](https://www.mathsisfun.com/numbers/images/pascals-triangle-doubles.svg)

The values of the triangle are the values for n choose r, where the rows are n from 0 to 8, and from left to right of each row is r where 0 $\le$ r $\le$ n. 

$${0 \choose 0}$$
$${1 \choose 0} {1 \choose 1}$$
$${2 \choose 0} {2 \choose 1} {2 \choose 2}$$
$${3 \choose 0} {3 \choose 1} {3 \choose 2} {3 \choose 3}$$
$$...$$
$$...$$
$$...$$
$${8 \choose 0} {8 \choose 1} {8 \choose 2} {8 \choose 3} {8 \choose 4} {8 \choose 5} {8 \choose 6} {8 \choose 7} {8 \choose 8}$$

We can build this in python and input N. 

In [22]:
N = input("Input N, the number of things to choose from: ")
N = int(N)
n_list = []
for n in range(N+1):
    r_list = []
    for r in range(n+1):
        ncr = math.comb(n,r)
        r_list.append(ncr)
    n_list.append(r_list)
    
n_list

# Print Pascals Triangle https://www.pythonpool.com/pascals-triangle-python/
for i in range(len(n_list)):
    print("    "*(len(n_list)-i) ,end=" ",sep=" ")
    for j in range(0,i+1):
        print("{0:6}".format(n_list[i][j]),end=" ",sep=" ")
    print("        <------", sum(n_list[i]))
    print()

Input N, the number of things to choose from: 4
                          1         <------ 1

                      1      1         <------ 2

                  1      2      1         <------ 4

              1      3      3      1         <------ 8

          1      4      6      4      1         <------ 16



The sum of the rows double each time, i.e, they are powers of 2. 

The triangles first and last element of each row is always 1. This is because: $${n \choose 0} == {n \choose n}$$. 

The triangle is also symmetrical. We can see the numbers on the left match the numbers on the right. This is because: $${n \choose r} == {n \choose n-r}$$. 


<h3 style="color: #001a79;">Exercise 2</h3>

<hr style="border-top: 1px solid #001a79;" />

Note that there are the same number of ways to get 4 tails as there to get 4 heads. Explain why this is.

The general formula is: $${n \choose r} = {n! \over r!(n-r)!}$$

where n is the number of things to choose from, and we choose r of them. 

https://www.mathsisfun.com/combinatorics/combinations-permutations.html

n is the number of coins and r is the number of tails. 

In [ ]:
r = 4
n = 10

In [ ]:
import math
combinations = (math.factorial(n)) // ((math.factorial(r)) * (math.factorial(n - r)))
combinations

This returns the same number of combinations if we set r to be the number of heads. 

It doesnt matter whether it is (4 heads and 6 tails) or (4 tails and 6 heads) the number of combinations will be 210.  

In [ ]:
n = 10
r = [x for x in range(11)]

In [ ]:
import math
combinations = []
for x in r:
    comb = (math.factorial(n)) // ((math.factorial(x)) * (math.factorial(n - x)))
    combinations.append(comb)
combinations

In [ ]:
import matplotlib.pyplot as plt
x = ["0T 10H", 
"1T 9H",
"2T 8H",
"3T 7H",
"4T 6H",
"5T 5H",
"6T 4H",
"7T 3H",
"8T 2H",
"9T 1H",
"10T 0H"]

# function to add value labels https://www.geeksforgeeks.org/adding-value-labels-on-a-matplotlib-bar-chart/
def addlabels(x,y):
    for i in range(len(x)):
        plt.text(i, y[i], y[i], ha = 'center')

plt.figure(figsize = (10, 5))
addlabels(x, combinations)
barlist = plt.bar(x, combinations)

# https://stackoverflow.com/a/18973430
barlist[4].set_color('r')
barlist[6].set_color('r')